From ccac9c6a1010f162b29d773f785b6a2b16daf2b7 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Tue, 31 May 2011 16:06:54 -0600
Subject: [PATCH 103/177] - ignore version in namespace change operations,
 always apply them; - do not modify ctime on link, unlink, rename, this is
 done by consequent setattr fop from teh log.

---
 mdservice/md_foms.c | 90 ++++++++++++++++++++++-------------------------------
 mdstore/mdstore.c   |  4 +--
 2 files changed, 39 insertions(+), 55 deletions(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index 6a28834..55dc504 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -273,18 +273,16 @@ static int c2_md_link_fom_state(struct c2_fom *fom)
         rc = c2_md_store_locate(site->s_mdstore, &tfid, &cob, 
                                 C2_MD_STORE_LOCATE_STORED,
                                 &ctx->fc_tx->tx_dbtx);
-        if (rc == 0) {
-                if (req->l_body.b_version > cob->co_nsrec.cnr_version) {
-                        name = c2_bitstring_alloc(req->l_name.n_name,
-                                                  req->l_name.n_len);
-                        rc = c2_md_store_link(site->s_mdstore, &pfid,
-                                              cob, name, 
-                                              &ctx->fc_tx->tx_dbtx);
-                        c2_bitstring_free(name);
-                }
-        } else {
+        if (rc)
                 goto out;
-        }
+
+        name = c2_bitstring_alloc(req->l_name.n_name,
+                                  req->l_name.n_len);
+        rc = c2_md_store_link(site->s_mdstore, &pfid,
+                              cob, name, 
+                              &ctx->fc_tx->tx_dbtx);
+        c2_bitstring_free(name);
+
         c2_cob_put(cob);
         if (rc == 0) {
                 svc = fom->fo_fop_ctx->fc_service;
@@ -370,14 +368,12 @@ static int c2_md_unlink_fom_state(struct c2_fom *fom)
                 if (rc)
                         goto out;
 
-                if (req->u_body.b_version > cob->co_nsrec.cnr_version) {
-                        name = c2_bitstring_alloc(req->u_name.n_name,
-                                                  req->u_name.n_len);
-                        rc = c2_md_store_unlink(site->s_mdstore, &pfid,
-                                                cob, name, 
-                                                &ctx->fc_tx->tx_dbtx);
-                        c2_bitstring_free(name);
-                }
+                name = c2_bitstring_alloc(req->u_name.n_name,
+                                          req->u_name.n_len);
+                rc = c2_md_store_unlink(site->s_mdstore, &pfid,
+                                        cob, name, 
+                                        &ctx->fc_tx->tx_dbtx);
+                c2_bitstring_free(name);
                 c2_cob_put(cob);
         }
         if (rc == 0) {
@@ -406,7 +402,8 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         struct c2_fid             tfid;
         struct c2_fid             pfid_tgt;
         struct c2_fid             pfid_src;
-        struct c2_cob_attr        attr;
+        struct c2_bitstring      *name_src;
+        struct c2_bitstring      *name_tgt;
         struct c2_service        *svc;
         int                       rc;
 
@@ -429,7 +426,6 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         sbody = &req->r_sbody;
         tbody = &req->r_tbody;
 
-        c2_md_fop_cob2attr(&attr, sbody);
         c2_md_make_fid(&pfid_src, &sbody->b_pfid);
         c2_md_make_fid(&pfid_tgt, &tbody->b_pfid);
         c2_md_make_fid(&tfid, &sbody->b_tfid);
@@ -440,43 +436,31 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         if (rc)
                 goto out;
 
-        if (sbody->b_version > cob->co_nsrec.cnr_version) {
-                struct c2_bitstring *name_src;
-                struct c2_bitstring *name_tgt;
                 
-                name_src = c2_bitstring_alloc(req->r_sname.n_name, 
-                                              req->r_sname.n_len);
-                if (!name_src) {
-                        c2_cob_put(cob);
-                        rc = -ENOMEM;
-                        goto out;
-                }
+        name_src = c2_bitstring_alloc(req->r_sname.n_name, 
+                                      req->r_sname.n_len);
+        if (!name_src) {
+                c2_cob_put(cob);
+                rc = -ENOMEM;
+                goto out;
+        }
                 
-                name_tgt = c2_bitstring_alloc(req->r_tname.n_name, 
-                                              req->r_tname.n_len);
-                if (!name_tgt) {
-                        c2_bitstring_free(name_src);
-                        c2_cob_put(cob);
-                        rc = -ENOMEM;
-                        goto out;
-                }
+        name_tgt = c2_bitstring_alloc(req->r_tname.n_name, 
+                                      req->r_tname.n_len);
+        if (!name_tgt) {
+                rc = -ENOMEM;
+                goto out_free_name_src;
+        }
 
-                rc = c2_md_store_rename(site->s_mdstore, &pfid_tgt, 
-                                        &pfid_src, cob, name_tgt, 
-                                        name_src, &ctx->fc_tx->tx_dbtx);
-                c2_bitstring_free(name_src);
-                c2_bitstring_free(name_tgt);
+        rc = c2_md_store_rename(site->s_mdstore, &pfid_tgt, 
+                                &pfid_src, cob, name_tgt, 
+                                name_src, &ctx->fc_tx->tx_dbtx);
+        c2_bitstring_free(name_tgt);
 
-                /*
-                 * Let's update attributes given we found that sbody contains
-                 * new set of then. Most of all we need to update version.
-                 */
-                if (rc == 0) {
-                        rc = c2_md_store_setattr(site->s_mdstore, cob,
-                                                 &attr, &ctx->fc_tx->tx_dbtx);
-                }
-        }
+out_free_name_src:
+        c2_bitstring_free(name_src);
         c2_cob_put(cob);
+
         if (rc == 0) {
                 svc = fom->fo_fop_ctx->fc_service;
 	        svc->s_ops->so_reply_post(svc, fop_rep, ctx->fc_cookie);
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index fe7ab71..b7fd612 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -372,9 +372,9 @@ int c2_md_store_rename(struct c2_md_store       *md,
         if (rc)
                 goto out;
 
+#if 0
         cob->co_nsrec.cnr_ctime = now;
 
-#if 0
         if (c2_fid_eq(pfid_src, pfid_tgt)) {
                 /*
                  * Update mtime in parent.
@@ -428,12 +428,12 @@ int c2_md_store_rename(struct c2_md_store       *md,
                 if (rc)
                         goto out;
         }
-#endif
 
         /*
          * Update cob on store to save changed cntr and ctime.
          */
         rc = c2_cob_update(cob, &cob->co_nsrec, NULL, NULL, tx);
+#endif
 out:
         c2_free(srckey);
         c2_free(tgtkey);
-- 
1.8.3.2

